File: src/main/java/org/nwapw/abacus/tree/Reducer.java
package org.nwapw.abacus.tree;

/**
 * Interface used to reduce a tree into a single value.
 *
 * @param <T> the value to reduce into.
 */
public interface Reducer<T> {

    /**
     * Reduces the given tree into a single value of type T.
     *
     * @param node     the node being passed in to be reduced.
     * @param children the already-reduced children of this node.
     * @return the resulting value from the reduce.
     */
    public T reduceNode(TreeNode node, Object... children);

}

File: core/src/main/kotlin/org/nwapw/abacus/tree/Reducer.kt
package org.nwapw.abacus.tree

/**
 * Reducer interface that takes a tree and returns a single value.
 *
 * The reducer walks the tree, visiting the children first, converting them into
 * a value, and then attempts to reduce the parent. Eventually, the single final value is returned.
 */
interface Reducer<out T> {

    /**
     * Reduces the given tree node, given its already reduced children.
     *
     * @param treeNode the tree node to reduce.
     * @param children the list of children, of type T.
     */
    fun reduceNode(treeNode: TreeNode, vararg children: Any): T?

}